mineinabyss-features/src/main/kotlin/com/mineinabyss/relics/StarCompassSystem.kt

```diff
@@ -13,58 +11,68 @@ import com.mineinabyss.geary.papermc.access.toGeary
+import com.mineinabyss.idofront.time.ticks
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
 import org.bukkit.Material
 import org.bukkit.entity.Player
-import org.bukkit.inventory.ItemStack
+import org.bukkit.inventory.ItemFlag
 import org.bukkit.inventory.meta.CompassMeta
-import kotlin.time.Duration.Companion.seconds
 
-class StarCompassSystem : RepeatingSystem(interval = 0.1.seconds) {
-    private val TargetScope.starCompass by get<StarCompass>()
-    private val TargetScope.item by get<ItemStack>()
-
-    override fun TargetScope.tick() {
-        val player = entity.parent?.get<Player>() ?: return
-        val playerBar = player.toGeary().getOrSetPersisting { PlayerCompassBar() }
-        val sectionCenter = player.location.section?.region?.center
+@Serializable
+@SerialName("mineinabyss:toggle_starcompass_hud")
+class ToggleDepthHud
 
-        if (sectionCenter != null) starCompass.compassLocation =
-            Location(player.world, sectionCenter.x.toDouble(), 0.0, sectionCenter.z.toDouble())
-        else starCompass.compassLocation = null
+class ToggleStarCompassHudSystem : GearyListener() {
+    private val TargetScope.player by get<Player>()
+    private val SourceScope.starCompass by get<StarCompass>()
+    private val EventScope.hasStarCompass by family { has<ToggleDepthHud>() }
 
-        // Let player toggle between having a bossbar-compass or item compass
-        if (player.toGeary().has<HideBossBarCompass>()) {
+    @Handler
+    fun TargetScope.toggleDepth(source: SourceScope) {
+        val item = player.inventory.itemInMainHand
+        if (player.toGeary().has<ShowStarCompassHud>()) {
             item.type = Material.COMPASS
             item.editItemMeta {
                 this as CompassMeta
-                lodestone = starCompass.compassLocation
+                lodestone = player.toGeary().get<ShowStarCompassHud>()?.lastSection?.getSectionCenter()
                 isLodestoneTracked = false
+                itemFlags.add(ItemFlag.HIDE_ENCHANTS)
             }
-            player.hideBossBar(playerBar.compassBar)
-        } else {
+            player.toGeary().remove<ShowStarCompassHud>()
+        }
+        else {
             item.type = Material.PAPER
-            player.bossbarCompass(starCompass.compassLocation, playerBar.compassBar)
+            item.removeItemFlags(ItemFlag.HIDE_ENCHANTS)
+            player.toGeary().setPersisting(ShowStarCompassHud(player.location.section))
         }
     }
 }
 
-//TODO fix when component removal events get implemented
-@AutoScan
-class RemoveStarCompassBar : GearyListener() {
+class StarCompassSystem(private val feature: RelicsFeature) : RepeatingSystem(interval = 2.ticks) {
+    private val TargetScope.starCompass by get<StarCompass>()
+
+    override fun TargetScope.tick() {
+        val player = entity.parent?.get<Player>() ?: return
+        player.toggleHud(feature.starcompassHudId, player.toGeary().has<ShowStarCompassHud>())
```

Comment:
This kind of task should never be done in a repeating system, you care about reacting to updates to components, nothing here must be done every tick, it's incredibly inefficient. Use an event listener.

Reply:
iirc geary didnt have anything to check when a component was added?
think we talked about this and was told to use this for now, did it change?

Reply:
Geary is missing an onRemoved call so I was unable to do this via a geary listener
Changed it now to toggle in the function itself so no need for a separate listener, atleast right now